### ISSUE TITLE
The printf family of functions doesn't pad numbers properly when the minus flag is present

### ISSUE THREAD

**Issue by GabrielRavier** [jart/cosmopolitan#772]
The following program:

```cpp
#include <stdio.h>

int main()
{
    printf("<%-4.3d>\n", 42);
}
```

outputs `<042 >` on the 4 libcs I have tested it on, whereas cosmopolitan instead outputs `<42  >`. This appears to be caused by cosmopolitan assuming that the presence of the `-` flag means printed numbers never need to be padded with 0s (see the code at the beginning of `__fmt_ntoa_format`), even though the C standard specifies as much:

> The precision specifies the minimum number of digits to appear; if the value being converted can be represented in fewer digits, **it is expanded with leading zeros**.

...and there is no indication anywhere that the minus flag changes this in any way.

**Comment by GabrielRavier** [comment]
I've written a patch that *should* fix this (I'm not fully confident of it but it fixes this specific issue and passes cosmopolitan's tests along with the libc testsuiite I'm writing (in so far as it doesn't cause any regressions and fixes the tests that were failing due to this)).

```patch
From b7873d80eee6243d7c56b42838366d13157e5876 Mon Sep 17 00:00:00 2001
From: Gabriel Ravier <gabravier@gmail.com>
Date: Thu, 23 Mar 2023 18:22:05 +0100
Subject: [PATCH] Fix padding+minus flag on numbers for printf-family functions

The C standard states, for conversions using the d, i, b, B, o, u, x or X conversion specifiers:
> The precision specifies the minimum number of digits to appear; if
> the value being converted can be represented in fewer digits, it is
> expanded with leading zeros.
- C standard, 7.23.6.1. The fprintf function

However, cosmopolitan currently suppresses the addition of leading
zeros when the minus flag is set. This is not reflected by anything
within the C standard, meaning that behavior is incorrect.

This patch fixes this.
---
 libc/fmt/ntoa.c | 21 ++++++++++-----------
 1 file changed, 10 insertions(+), 11 deletions(-)

diff --git a/libc/fmt/ntoa.c b/libc/fmt/ntoa.c
index 8bf081768..116a9ed7a 100644
--- a/libc/fmt/ntoa.c
+++ b/libc/fmt/ntoa.c
@@ -34,17 +34,16 @@ static int __fmt_ntoa_format(int out(const char *, void *, size_t), void *arg,
                              unsigned char flags) {
   unsigned i;
   /* pad leading zeros */
-  if (!(flags & FLAGS_LEFT)) {
-    if (width && (flags & FLAGS_ZEROPAD) &&
-        (negative || (flags & (FLAGS_PLUS | FLAGS_SPACE)))) {
-      width--;
-    }
-    while ((len < prec) && (len < BUFFER_SIZE)) {
-      buf[len++] = '0';
-    }
-    while ((flags & FLAGS_ZEROPAD) && (len < width) && (len < BUFFER_SIZE)) {
-      buf[len++] = '0';
-    }
+  if (width && (flags & FLAGS_ZEROPAD) && !(flags & FLAGS_LEFT) &&
+      (negative || (flags & (FLAGS_PLUS | FLAGS_SPACE)))) {
+    width--;
+  }
+  while ((len < prec) && (len < BUFFER_SIZE)) {
+    buf[len++] = '0';
+  }
+  while ((flags & FLAGS_ZEROPAD) && !(flags & FLAGS_LEFT) && (len < width) &&
+         (len < BUFFER_SIZE)) {
+    buf[len++] = '0';
   }
   /* handle hash */
   if (flags & FLAGS_HASH) {
-- 
2.39.2

```

**Comment by GabrielRavier** [comment]
Here's another version of that patch, I think that one's cleaner in how it handles any combination of the 0 and - flags (it also makes far more sense to me and seems quite likely to be correct now):

```patch
From 0901605dc199a0f1494cda762df33b532ff34d13 Mon Sep 17 00:00:00 2001
From: Gabriel Ravier <gabravier@gmail.com>
Date: Thu, 23 Mar 2023 18:22:05 +0100
Subject: [PATCH] Fix padding+minus flag on numbers for printf-family functions

The C standard states, for conversions using the d, i, b, B, o, u, x or X conversion specifiers:
> The precision specifies the minimum number of digits to appear; if
> the value being converted can be represented in fewer digits, it is
> expanded with leading zeros.
- C standard, 7.23.6.1. The fprintf function

However, cosmopolitan currently suppresses the addition of leading
zeros when the minus flag is set. This is not reflected by anything
within the C standard, meaning that behavior is incorrect.

This patch fixes this.
---
 libc/fmt/ntoa.c | 24 +++++++++++-------------
 1 file changed, 11 insertions(+), 13 deletions(-)

diff --git a/libc/fmt/ntoa.c b/libc/fmt/ntoa.c
index 8bf081768..8b1878df0 100644
--- a/libc/fmt/ntoa.c
+++ b/libc/fmt/ntoa.c
@@ -34,17 +34,15 @@ static int __fmt_ntoa_format(int out(const char *, void *, size_t), void *arg,
                              unsigned char flags) {
   unsigned i;
   /* pad leading zeros */
-  if (!(flags & FLAGS_LEFT)) {
-    if (width && (flags & FLAGS_ZEROPAD) &&
-        (negative || (flags & (FLAGS_PLUS | FLAGS_SPACE)))) {
-      width--;
-    }
-    while ((len < prec) && (len < BUFFER_SIZE)) {
-      buf[len++] = '0';
-    }
-    while ((flags & FLAGS_ZEROPAD) && (len < width) && (len < BUFFER_SIZE)) {
-      buf[len++] = '0';
-    }
+  if (width && (flags & FLAGS_ZEROPAD) &&
+      (negative || (flags & (FLAGS_PLUS | FLAGS_SPACE)))) {
+    width--;
+  }
+  while ((len < prec) && (len < BUFFER_SIZE)) {
+    buf[len++] = '0';
+  }
+  while ((flags & FLAGS_ZEROPAD) && (len < width) && (len < BUFFER_SIZE)) {
+    buf[len++] = '0';
   }
   /* handle hash */
   if (flags & FLAGS_HASH) {
@@ -134,8 +132,8 @@ int __fmt_ntoa(int out(const char *, void *, size_t), void *arg, va_list va,
   bool neg;
   uint128_t value, sign;
 
-  /* ignore '0' flag when prec is given */
-  if (flags & FLAGS_PRECISION) {
+  /* ignore '0' flag when prec or minus flag is given */
+  if (flags & (FLAGS_PRECISION | FLAGS_LEFT)) {
     flags &= ~FLAGS_ZEROPAD;
   }
 
-- 
2.39.2

```

**Comment by G4Vi** [comment]
Fixed by #787.
